FF116 Audio Output Device API updates (#27818)
mdn/content: files/en-us/web/api/mediadevices/selectaudiooutput/index.md

{{APIRef("WebRTC")}} {{SeeCompatTable}}

> The {{domxref("MediaDevices.selectAudioOutput()")}} method of the [Web Audio Output Devices API](/en-US/docs/Web/API/Audio_Output_Devices_API) prompts the user to select an audio output device, such as a speaker or headset, and grants user-permission to use the selected device as an audio output sink.

Reviewer: ```suggestion
The {{domxref("MediaDevices.selectAudioOutput()")}} method of the [Audio Output Devices API](/en-US/docs/Web/API/Audio_Output_Devices_API) prompts the user to select an audio output device, such as a speaker or headset. If the user selects a device, the method grants user permission to use the selected device as an audio output sink.
```

Author: Yes, this is much better.

Following selection, if the device is available it can be enumerated using [`MediaDevices.enumerateDevices()`](/en-US/docs/Web/API/MediaDevices/enumerateDevices) and set as the audio output sink using [`HTMLMediaElement.setSinkId()`](/en-US/docs/Web/API/HTMLMediaElement/setSinkId).

> The method can only be called in a [secure context](/en-US/docs/Web/Security/Secure_Contexts) with [transient user activation](/en-US/docs/Web/Security/User_activation), and will only offer audio output devices if allowed by the page [`Permission-Policy: speaker-selection`](/en-US/docs/Web/HTTP/Headers/Permissions-Policy/speaker-selection).

Reviewer: I would omit this paragraph, as it's covered below.

Author: I thought about this for a bit - it's my main problem with a security section - core information that isn't part of the main description. That's why we need some kind of "there are security considerations [see link]" banner.

I will remove though.

    - `deviceId` {{Optional_Inline}}

      - : A string representing the ID of a single previously exposed/permitted device.
        If not set, a prompt with all available audio output devices will be displayed.

        The option is intended for applications that want to store a device id so that the same device can be used by default in future sessions.

>         Note that the method may return a new ID for the same device, and that persisted ids _must be passed_ through `selectAudioOutput()` successfully before they will work with {{domxref("HTMLMediaElement.setSinkId","setSinkId()")}}.

Author: FYI @wbamberg This is updated explanation of what it means to persist a device id

    - `deviceId` {{Optional_Inline}}

>       - : A string representing the id of the (only) device to display in the prompt (with default value: "").

Author: I had this wrong. It has no default value. It is used to set a single value if you want to cycle an Id you got previously for reselection.

>       - : A string representing the id of a single device to display in the prompt.

Reviewer 2: Perhaps "the id of a single previously exposed device."
Firefox will usually not prompt if permission has previously been granted for this device. if the permission has been removed, however, Firefox will prompt with all available devices, but will have the device indicated by `deviceId` preselected.

Author: Thanks @karlt .

```suggestion
      - : A string representing the id of a single previously exposed/permitted device.
```

>         If not set, all audio output devices for which permission is not explicitly denied will be displayed.

Reviewer 2: Firefox doesn't doing any blocking of individual devices and this feature is not implemented elsewhere, so I'd remove "for which permission is not explicitly denied".
This could say "all available audio output devices will be displayed" or "a prompt with all available audio output devices will be displayed".

Author: The spec seemed to indicate that if the permission policy denied access the prompt would not include devices of the specified type other than the default. My assumption was that there might still be devices in that list - if I'm wrong then absolutely!

```suggestion
        If not set, a prompt with all available audio output devices will be displayed.
```

Reviewer 2: I'm not clear which part of the spec seemed to indicate only a default device.  `enumerateDevices()` behaves like that.

https://www.w3.org/TR/permissions/#dfn-prompt-the-user-to-choose is vague and user agents could do a variety of things.  I think "all available" is sufficient to capture that.

If "speaker-selection" feature/permissions policy is not available to the document, then ['If descriptor's permission state is "denied", reject p with a new DOMException whose name attribute has the value NotAllowedError, and abort these steps.'](https://w3c.github.io/mediacapture-output/#dom-mediadevices-selectaudiooutput) takes effect and there will be no prompt.

        The option is intended for applications that want to use persisted device ids.

        > **Note:** A user agent may choose to skip prompting the user if a specified non-null id was previously exposed to the user by `selectAudioOutput()` in an earlier session.

>         > In this case the user agent may simply resolve with this device id, or a new id for the same device if it has changed.

Author: This is just to make it clear that what Firefox actually does is a valid alternative. We try not to say too much about what specific implementations do other than in the compatibility data, and only then when it is a compatibility "issue". 

```suggestion
        > In this case the user agent may simply resolve with this device id, or a new id for the same device if it has changed.
        > If permission for the specified device was previously granted but has since been revoked, the user-agent might display all allowed devices, highlighting the one with the specified ID.
```

Reviewer 2: Yes, this is not a common case and I did not mean to imply that it should be added, but yes it might be helpful to have this or something else so readers know that they might get a different device back.

    - `deviceId` {{Optional_Inline}}

>       - : A string representing the id of a single previously exposed/permitted device.

Reviewer: We use "ID" in the other page.

Author: ```suggestion
      - : A string representing the ID of a single previously exposed/permitted device.
```

        If not set, a prompt with all available audio output devices will be displayed.

>         The option is intended for applications that want to use persisted device ids.

Reviewer: I'm not sure what a persisted device ID is. Do you mean, the application stored the ID it used last time and wants to use the same device again? If so I think it is worth spelling this out a bit more.

Author: Yes. Will do.

        > **Note:** A user agent may choose to skip prompting the user if a specified non-null id was previously exposed to the user by `selectAudioOutput()` in an earlier session.
        > In this case the user agent may simply resolve with this device id, or a new id for the same device if it has changed.
        > If permission for the specified device was previously granted but has since been revoked, the user-agent might display all allowed devices, highlighting the one with the specified ID.
        >
        > Note that persisted ids _must be passed_ through `selectAudioOutput()` successfully before they will work with {{domxref("HTMLMediaElement.setSinkId","setSinkId()")}}.

### Return value

> 

Reviewer: Again "fulfills with" is more common than "receives" in our docs AFAIK, and again it would be good to consider how to describe exceptions.

Author: Thanks. Fixed.
